### PR TITLE
[SMALLFIX] generalize extension factory registry

### PR DESCRIPTION
**Pull request by bf8086**
Extract extension loading logic from UnderFileSystemFactoryRegistry and put it in a reusable generic class.

### CODE REVIEW

File: core/common/src/main/java/alluxio/extensions/ExtensionFactory.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.extensions;
+
+import javax.annotation.Nullable;
+
+public interface ExtensionFactory<T, S> {
```
Comment (gpang) [inline core/common/src/main/java/alluxio/extensions/ExtensionFactory.java]: comment on this? What are `T` and `S`?

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java [hunk 2]
```diff
@@ -44,58 +28,20 @@
...
-   * creation occurs.
-   */
-  private static final List<UnderFileSystemFactory> FACTORIES = new CopyOnWriteArrayList<>();
-
-  public static final String LIB_DIR =
-      PathUtils.concatPath(Configuration.get(PropertyKey.HOME), "lib");
-
-  private static boolean sInit = false;
+  private static final String EXTENSION_PATTERN = "alluxio-underfs-*.jar";
+  private static ExtensionFactoryRegistry<UnderFileSystemFactory, UnderFileSystemConfiguration>
```
Comment (gpang) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: how are you supposed to use `ExtensionFactoryRegistry`? extend it or create one, or create a static one?

**Comment by gpang** [comment]
@madanadit Can you take a look at this?

File: core/common/src/main/java/alluxio/extensions/ExtensionFactory.java (hunk 1 above)
Comment (bf8086) [reply core/common/src/main/java/alluxio/extensions/ExtensionFactory.java]: Added.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java (hunk 2 above)
Comment (bf8086) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: Normally you can just create one and hold on to it. For `UnderFileSystemFactory` there are many integration tests reply on the registry being static to register test factories, so we have to keep it static.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java [hunk 3]
```diff
@@ -44,58 +28,20 @@
...
-   * registered by tests. All other UFS factories will be discovered and service loaded when ufs
-   * creation occurs.
-   */
-  private static final List<UnderFileSystemFactory> FACTORIES = new CopyOnWriteArrayList<>();
-
-  public static final String LIB_DIR =
-      PathUtils.concatPath(Configuration.get(PropertyKey.HOME), "lib");
-
-  private static boolean sInit = false;
+  private static final String EXTENSION_PATTERN = "alluxio-underfs-*.jar";
```
Comment (madanadit) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: not: rename to `UFS_EXTENSION_PATTERN`?

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java [hunk 4]
```diff
@@ -44,58 +28,20 @@
...
+  private static ExtensionFactoryRegistry<UnderFileSystemFactory, UnderFileSystemConfiguration>
+      sInstance;
```
Comment (madanadit) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: nit: rename to `sRegistryInstance`?

**Comment by bf8086** [comment]
@gpang @madanadit @calvinjia PTAL.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java [hunk 5]
```diff
@@ -155,101 +101,14 @@ public static UnderFileSystemFactory find(
...
-        for (UnderFileSystemFactory factory : extensionServiceLoader) {
-          LOG.debug("Discovered an Under File System Factory implementation {} - {} in jar {}",
-              factory.getClass(), factory.toString(), jarPath);
-          register(factory, factories);
-        }
-      } catch (Throwable t) {
-        LOG.warn("Failed to load jar {}: {}", jar, t.getMessage());
-      }
-    }
+    return sRegistryInstance.findAll(path, ufsConf);
```
Comment (madanadit) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: It seems safer to wrap `sRegistryInstance` in a getInstance method to protect against null references https://crunchify.com/thread-safe-and-a-fast-singleton-implementation-in-java/

Comment (gpang) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: This sounds like a good idea.

File: core/common/src/main/java/alluxio/extensions/ExtensionFactory.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public interface ExtensionFactory<T, S> {
+  /**
+   * Creates a new extension for the given path. An {@link IllegalArgumentException} is
+   * thrown if this factory does not support extension for the given path or if the configuration
+   * provided is insufficient to create a client.
+   *
+   * @param path file path with scheme for which the extension will be created
+   * @param conf optional configuration object for the extension, may be null
+   * @return the client
```
[the remark this answers is not in the document]
Comment (bf8086) [reply core/common/src/main/java/alluxio/extensions/ExtensionFactory.java]: Updated.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java (hunk 5 above)
Comment (bf8086) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java]: The static initializer of the class calls to the `init()` method to ensure the instance is created before anyone can access it. This is unchanged in this PR from how we originally initialize resources with static functions.

**Pull request merged**